### ISSUE TITLE
Long running application sometimes stops to get watched DBUS changes

### ISSUE THREAD

**Issue by AdrianEggenberger** [tmds/Tmds.DBus#164]
Hi

We are using the release 0.10 on a linux system to monitor the NetworkManager. During startup of our application we setup a Watch instance for changes on the wired network as seen in my code snippet below.

This works as expected normally, but we see cases that changes on the wired device or wired interface are never received in the handler methods. It seems as the connection to the DBUS is dead in that case.

What could cause to this behavior? Is there a connection loss that can happen (and detected)? Is there limitation we could be running into?

Regards Adrian

```
// create network manager
_networkManager = _dbusConnection.CreateProxy<INetworkManager>(networkManagerService, networkManagerObjectPath);

// create network manager settings
_networkManagerSettings = _dbusConnection.CreateProxy<ISettings>(networkManagerService, new ObjectPath("/org/freedesktop/NetworkManager/Settings"));

// search devices
var devices = await _networkManager.GetDevicesAsync();
foreach (var device in devices)
{
	// get device type
	// https://people.freedesktop.org/~lkundrak/nm-docs/nm-dbus-types.html#NMDeviceType
	var deviceType = await device.GetDeviceTypeAsync();

	if (deviceType == 1) // NM_DEVICE_TYPE_ETHERNET
	{
		_wiredDevice = device;
	}
}

// initialize devices
if (_wiredDevice != null)
{
	_wiredDeviceWatch = await _wiredDevice.WatchPropertiesAsync(WiredDevicePropertiesChangedHandler);

	_wiredInterface = _dbusConnection.CreateProxy<IWired>(networkManagerService, _wiredDevice.ObjectPath);
	_wiredInterfaceWatch =
		await _wiredInterface.WatchPropertiesAsync(WiredInterfacePropertiesChangedHandler);

	var wiredProperties = await _wiredInterface.GetAllAsync();
	Log.Info($"Ethernet Mac-Address: {wiredProperties.HwAddress}");
}
```

**Comment by tmds** [comment]
> What could cause to this behavior? Is there a connection loss that can happen (and detected)? Is there limitation we could be running into?

You can add a `Action<Exception>` argument to a signal method that will be called when the connection is closed.
The Connection class itself also emits state changes when it gets closed.

If you attach a debugger to your app and look at the `Connection` instance, probably it will give you an idea of its state and help debug the issue further.

**Comment by madewokherd** [comment]
From print statement debugging, it seems that sometimes when TransportSocket.DataAvailable gets EAGAIN, and ReadAsync completes the Task, ReadCountAsync never resumes from its await of ReadAsync. I'm not sure how that can happen, but I also don't really understand yet how this is supposed to work.

**Comment by madewokherd** [comment]
I have this as a work-around.
```diff
\diff --git a/src/Tmds.DBus/Transports/TransportSocket.cs b/src/Tmds.DBus/Transports/TransportSocket.cs
index 6432a0f..f6db7df 100644
--- a/src/Tmds.DBus/Transports/TransportSocket.cs
+++ b/src/Tmds.DBus/Transports/TransportSocket.cs
@@ -142,7 +142,19 @@ private void DataAvailable(object sender, SocketAsyncEventArgs e)
                 int errno = -rv;
                 if (errno == EAGAIN)
                 {
-                    ReadAsync(readContext.Buffer, readContext.Offset, readContext.Count, readContext.FileDescriptors);
+                    var task = ReadAsync(readContext.Buffer, readContext.Offset, readContext.Count, readContext.FileDescriptors);
+                    if (task != tcs.Task)
+                    {
+                        readContext.Tcs = null;
+                        if (task.IsFaulted)
+                        {
+                            tcs.SetException(task.Exception);
+                        }
+                        else
+                        {
+                            tcs.SetResult(task.Result);
+                        }
+                    }
                 }
                 else
                 {
```
I think it would be better for ReadAsync to always set its result directly on readContext.Tcs, but I can't seem to get this right.

**Comment by madewokherd** [comment]
Actually, why have DataAvailable attempt to read by itself at all? Couldn't it just delegate to ReadAsync in all cases?

**Comment by AdrianEggenberger** [comment]
@madewokherd That's an interesting finding. I'll test your patch in my project and get back with feedback if it fixes the issue also here.

**Comment by tmds** [comment]
```
if (task != tcs.Task)
```
This condition shouldn't be possible.

This is how it is supposed to work. (For your connection, `_supportsFdPassing` is `true`.)

The user calls `ReadAsync`. When the Socket has data, `DataAvailable` gets called.
In `DataAvailable`, we try to read the data by calling `DoRead`.
When `DoRead` returns `EAGAIN`, we call `ReadAsync`.

Since we haven't set `readContext.Tcs` to `null`, we'll keep the `TaskCompletionSource` that was created when the user called `ReadAsync` initially:

https://github.com/tmds/Tmds.DBus/blob/5f89abed6e5e7a052d442d58a427791b4f0dc56e/src/Tmds.DBus/Transports/TransportSocket.cs#L246-L247

**Comment by dmsch** [comment]
We had similar problem on some clients, interestingly it disappeared after updating from Ubuntu 20.04.4 5.11.0-27 to 5.13.0-52

**Comment by madewokherd** [comment]
Well, again based on print statement debugging, it sometimes executes this line when called by DataAvailable, which creates a new Task: https://github.com/tmds/Tmds.DBus/blob/60561919a1d1429063abf128d9f74348628ac01b/src/Tmds.DBus/Transports/TransportSocket.cs#L253

**Comment by madewokherd** [comment]
The problem is the assumption that DoRead will return the same thing to both methods. The call in ReadAsync happens later, so the amount of data available to read may have changed.

**Comment by AdrianEggenberger** [comment]
A short update to this topic from my side. I've integrated the from @madewokherd  in our solution 4 days ago. Since then the problem with loosing events never again occurred.

**Comment by tmds** [comment]
Great!

This bug has been there for 5 years.
The situation is unexpected. When a socket says it is ready for reading, the next read is expected to return data, not EAGAIN.

It would be interesting to know if this behavior comes from the kernel, or whether it comes from .NET.

@madewokherd can you make an `strace` that shows the syscalls that are being made?

Instead of `DataAvailable(null, null);`, you can call `Environment.FailFast` so the app stops when we get in the problematic case.

These are the calls I'm interested in:
```
strace -f -e epoll_ctl,epoll_wait,recvmsg,recv,recvfrom dotnet <yourapp.dll>
```

**Comment by madewokherd** [comment]
And right after I said that, I got the failure.

[strace.log.gz](https://github.com/tmds/Tmds.DBus/files/9086137/strace.log.gz)

**Comment by tmds** [comment]
The only think that stands out from the strace is that at the end there are two successive calls to `recvmsg` with `MSG_PEEK`.

The `MSG_PEEK` ones are those that `ReceiveAsync` makes. Those without are by `DoRead`.

Normally after `MSG_PEEK`that returns `EAGAIN` there should be one without `MSG_PEEK`.

I don't see how we get in that situation, and I won't break my head over it.

**Comment by tmds** [comment]
The fix is part of 0.11.0 which was just published to nuget.org.
